[RFC] Suppress exceptions from logger code

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Suppress exceptions from logger code

Detailed description / Documentation draft:
Some code does not expect from the logger code to throw, and this can
lead to the problems like:
- marking distributed batch as broken (in
  StorageDistributedDirectoryMonitor)
- some MergeTree code may also be affected.

So instead of throwing the exception, it will be logged to the stderr
(if it is possible).

azat committed:
Suppress exceptions from logger code

Some code does not expect from the logger code to throw, and this can
lead to the problems like:
- marking distributed batch as broken (in
  StorageDistributedDirectoryMonitor)
- some MergeTree code may also be affected.

So instead of throwing the exception, it will be logged to the stderr
(if it is possible).

azat committed:
Cleanup 01594_too_low_memory_limits (by using clientError)

azat committed:
Disable 01594_too_low_memory_limits under sanitizers

alesapin commented on base/loggers/OwnSplitChannel.cpp:
```diff
@@ -26,16 +27,52 @@ void OwnSplitChannel::log(const Poco::Message & msg)
+        /// Also note, that we cannot log the exception here, since this
+        /// will lead to recursion, using regular tryLogCurrentException().
+        /// but let's log it into the stderr at least.
+        std::string message = "Cannot add message to the log:\n";
```
But these lines also can lead to exceptions, for example, `bad_alloc`. However, it should be better to throw them than silently catch and ignore them....

azat replied:
Right, I though that logging uses `LockExceptionInThread`/`BlockerInThread` but it does not.

To me it sounds like it just should use `LockExceptionInThread`, since memory limit error during logging will hide log message (plus it is blocked in few places already, like `SystemLog<>::add`) @alesapin thoughts?

azat replied:
>To me it sounds like it just should use LockExceptionInThread,

Done

azat committed:
Block memory tracking during catching exceptions in OwnSplitChannel

nikitamikhaylov commented:
@Mergifyio update

alexey-milovidov commented on base/loggers/OwnSplitChannel.cpp:
```diff
@@ -26,16 +27,54 @@ void OwnSplitChannel::log(const Poco::Message & msg)
+
+        try
+        {
+            WriteBufferFromFileDescriptor out(STDERR_FILENO, 4096);
```
There is also `WriteBufferFromFileDescriptorDiscardOnFailure`

alexey-milovidov reviewed (approved):
Looks reasonable.

alexey-milovidov commented on base/loggers/OwnSplitChannel.cpp:
```diff
@@ -26,16 +27,54 @@ void OwnSplitChannel::log(const Poco::Message & msg)
+
+        try
+        {
+            WriteBufferFromFileDescriptor out(STDERR_FILENO, 4096);
```
We can also get rid of temporary allocation. We can write std::string contents directly.
Maybe write a loop with `write` and handle EINTR?

azat replied:
Yes, this looks better, done.

alexey-milovidov commented on base/loggers/OwnSplitChannel.cpp:
```diff
@@ -26,16 +27,54 @@ void OwnSplitChannel::log(const Poco::Message & msg)
+        }
+        catch (...)
+        {
+            /// Nothing can be done. Ignore the exception.
```
Then there will be no exceptions here.

azat committed:
Add writeRetry helper

azat committed:
Use writeRetry in case of exception during writing to the log

azat commented:
> [Docs check — Found 1 errors](https://clickhouse-test-reports.s3.yandex.net/24069/96242f3c2ce1bccfc3e9a2455c731d01ffc3a6e1/docs_check/docs_output.out.log)

```
ERROR: Cannot install -r requirements.txt (line 30) and urllib3>=1.26.5 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested urllib3>=1.26.5
    requests 2.24.0 depends on urllib3!=1.25.0, !=1.25.1, <1.26 and >=1.21.1
```

#24865

alexey-milovidov commented:
@mergifyio update

alexey-milovidov commented:
We need to make sure that the master is all green before merging this PR.

azat committed:
Revert "Cleanup 01594_too_low_memory_limits (by using clientError)"

--testmode can't be used since this disables logging while the
failed allocation comes from:

    2021.06.04 09:54:49.576479 [ 31070 ] {} <Error> ServerErrorHandler: Code: 241, e.displayText() = DB::Exception: Memory limit (for query) exceeded: would use 2.01 MiB (attempt to allocate chunk of 4096 bytes), maximum: 1.00 B, Stack trace (when copying this message, always include the lines below):

    0. /src/ch/clickhouse/.cmake/../contrib/libcxx/include/exception:133: std::exception::capture() @ 0xe2d88 in /src/ch/clickhouse/.cmake/programs/server/libclickhouse-server-libd.so
    1. /src/ch/clickhouse/.cmake/../contrib/libcxx/include/exception:111: std::exception::exception() @ 0x1e9b92 in /src/ch/clickhouse/.cmake/contrib/poco-cmake/Foundation/lib_poco_foundationd.so
    2. /src/ch/clickhouse/.cmake/../contrib/poco/Foundation/src/Exception.cpp:28: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1dd943 in /src/ch/clickhouse/.cmake/contrib/poco-cmake/Foundation/lib_poco_foundationd.so
    3. /src/ch/clickhouse/.cmake/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x2aaffd in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    4. /src/ch/clickhouse/.cmake/../src/Common/Exception.h:40: DB::Exception::Exception<char const*, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*&&, char const*&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) @ 0x2ebef0 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    5. /src/ch/clickhouse/.cmake/../src/Common/MemoryTracker.cpp:226: MemoryTracker::allocImpl(long, bool) @ 0x2eac82 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    6. /src/ch/clickhouse/.cmake/../src/Common/MemoryTracker.cpp:0: MemoryTracker::allocImpl(long, bool) @ 0x2eae3e in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    7. /src/ch/clickhouse/.cmake/../src/Common/CurrentMemoryTracker.cpp:47: CurrentMemoryTracker::(anonymous namespace)::allocImpl(long, bool) @ 0x27b621 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    8. /src/ch/clickhouse/.cmake/../src/Common/CurrentMemoryTracker.cpp:61: CurrentMemoryTracker::alloc(long) @ 0x27b541 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    9. /src/ch/clickhouse/.cmake/../src/Common/Allocator.h:96: Allocator<false, false>::alloc(unsigned long, unsigned long) @ 0x274b71 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    10. /src/ch/clickhouse/.cmake/../src/Common/PODArray.h:120: void DB::PODArrayBase<4ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::alloc<>(unsigned long) @ 0x305af3 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    11. /src/ch/clickhouse/.cmake/../src/Common/PODArray.h:146: void DB::PODArrayBase<4ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::realloc<>(unsigned long) @ 0x305cbb in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    12. /src/ch/clickhouse/.cmake/../src/Common/PODArray.h:182: void DB::PODArrayBase<4ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::reserveForNextSize<>() @ 0x306010 in /src/ch/clickhouse/.cmake/src/libclickhouse_common_iod.so
    13. /src/ch/clickhouse/.cmake/../src/Common/PODArray.h:0: void DB::PODArray<unsigned int, 4096ul, Allocator<false, false>, 15ul, 16ul>::push_back<unsigned int>(unsigned int&&) @ 0x5423a6a in /src/ch/clickhouse/.cmake/src/AggregateFunctions/libclickhouse_aggregate_functionsd.so
    14. /src/ch/clickhouse/.cmake/../src/Columns/ColumnVector.h:280: DB::ColumnVector<unsigned int>::insert(DB::Field const&) @ 0x3c6422 in /src/ch/clickhouse/.cmake/src/libclickhouse_columnsd.so
    15. /src/ch/clickhouse/.cmake/../base/loggers/OwnSplitChannel.cpp:60: DB::OwnSplitChannel::logSplit(Poco::Message const&) @ 0xa2833 in /src/ch/clickhouse/.cmake/base/loggers/libloggersd.so
    16. /src/ch/clickhouse/.cmake/../base/loggers/OwnSplitChannel.cpp:35: DB::OwnSplitChannel::log(Poco::Message const&) @ 0xa2588 in /src/ch/clickhouse/.cmake/base/loggers/libloggersd.so

This reverts commit 0cecc754709cc774c90269a1dc8653eac8725341.

azat committed:
Rewrite 01594_too_low_memory_limits

After #24483 had been merged the only place where the allocation may
fail is the insert into PODArray in DB::OwnSplitChannel::log, but this
MR ignores the errors in this function, so to check new behaviour
separate server is required.

azat commented on tests/queries/0_stateless/01594_too_low_memory_limits.sh:
```diff
@@ -1,16 +1,103 @@
-test $exit_code -eq 32 || exit 1
+# no sleep, since flushing to stderr should not be buffered.
+if ! grep -E -q 'Cannot add message to the log: Code: 60.*placeholder_table_name' clickhouse-server.stderr; then
+    echo "Adding message to the log should fail" >&2
```
I've even managed to write a test for this (while looking into the test failure).

alexey-milovidov commented:
Need to fix the test `01414_freeze_does_not_prevent_alters`.

azat commented:
>Need to fix the test 01414_freeze_does_not_prevent_alters.

@alexey-milovidov this is the real issue - #25025

azat commented:
@alexey-milovidov should I rebase this patch against upstream/master? Or passed checks is enough?

alexey-milovidov commented:
There are too many failed checks, let's merge just to make sure.
@mergifyio update

azat committed:
Do not check writeRetry errors in OwnSplitChannel

Should fix PVS warning:

  "There are identical sub-expressions '!writeRetry(2, "\n")' to the left and to the right of the '||' operator."

alexey-milovidov commented:
Something wrong with TestFlows.

vzakaznikov commented:
> We need to either remove TestFlows or to fix them @vzakaznikov
> Keeping flaky tests in master is not an option.

TestFlows check failed due to extended precision data type test. We will fix it.

```
    Assertion values
      assert that(snapshot("\n" + r.output.strip() + "\n", "tests", name=name, encoder=str)), error()
             ^ is = SnapshotError(
        filename=/ClickHouse/tests/testflows/extended_precision_data_types/snapshots/common.py.tests.snapshot
        snapshot_value="""

            log(toDecimal256(\'1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))	log(toDecimal256(\'-1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))
            172.69388197463743	nan
        """,
        actual_value="""

            log(toDecimal256(\'1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))	log(toDecimal256(\'-1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))
            172.69388197455342	nan
        """,
        diff="""
            --- /ClickHouse/tests/testflows/extended_precision_data_types/snapshots/common.py.tests.snapshot
            +++ 
            @@ -1,3 +1,3 @@

             log(toDecimal256(\'1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))	log(toDecimal256(\'-1000000000000000000000000000000000000000000000000000000000000000000000000000\', 0))
            -172.69388197463743	nan
            +172.69388197455342	nan
        """)
```

alexey-milovidov commented:
Ok. BTW, the least significant bit depends on the implementation that is selected dynamically depending on the availability of AVX2 on the machine.
